src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1251 @@
+		{
+			int latchFlags = WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH;
+			int rc = WaitLatch(MyLatch, latchFlags, (long) 1000, PG_WAIT_EXTENSION);
+
+			/* emergency bailout if postmaster has died */
+			if (rc & WL_POSTMASTER_DEATH)
+			{
+				proc_exit(1);
+			}
+
+			if (rc & WL_LATCH_SET)
+			{
+				ResetLatch(MyLatch);
+				CHECK_FOR_INTERRUPTS();
+			}
+
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, task, &hadError, &cancelled);
+
+			shm_mq_detach(responseq);
+		}
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		{
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, task, &hadError, &cancelled);
+
+			shm_mq_detach(responseq);
+		}
+
+		UNSET_NULLABLE_FIELD(task, pid);
+		task->status = BACKGROUND_TASK_STATUS_DONE;
+		if (hadError)
+		{
+			/*
+			 * When we had an error we need to decide if we want to retry (keep the
+			 * scheduled state), or move to failed state
+			 */
+			if (cancelled)
+			{
+				task->status = BACKGROUND_TASK_STATUS_CANCELLED;
+			}
+			else if (!task->retry_count)
+			{
+				SET_NULLABLE_FIELD(task, retry_count, 1);
+			}
+			else
+			{
+				(*task->retry_count)++;
+			}
+
+			if (task->retry_count)
+			{
+				if (*task->retry_count >= 3)
+				{
+					/* fail after 3 retries */
+					task->status = BACKGROUND_TASK_STATUS_ERROR;
+					UNSET_NULLABLE_FIELD(task, not_before);
+				}
+				else
+				{
+					/*
+					 * Per try we increase the delay as follows:
+					 *   retry 1: 1 min
+					 *   retry 2: 10 min
+					 *   retry 3: 60 min
+					 *
+					 * In the future we would like a callback on the job_type that could
+					 * distinguish the retry count and delay + potential jitter on a
+					 * job_type basis. For now we only assume this to be used by the
+					 * rebalancer and settled on the retry scheme above.
+					 */
+					int64 delayMs = 0;
+					switch (*(task->retry_count))
+					{
+						case 1:
+						{
+							delayMs = 1 * 60 * 1000;
+							break;
+						}
+
+						case 2:
+						{
+							delayMs = 10 * 60 * 1000;
+							break;
+						}
+
+						case 3:
+						default: /* just make sure retrying an hour for missed cases */
+						{
+							delayMs = 60 * 60 * 1000;
+							break;
+						}
+					}
+					TimestampTz notBefore = TimestampTzPlusMilliseconds(
+						GetCurrentTimestamp(), delayMs);
+					SET_NULLABLE_FIELD(task, not_before, notBefore);
+
+					task->status = BACKGROUND_TASK_STATUS_RUNNABLE;
+				}
+			}
+		}
+
+		UpdateBackgroundTask(task);
```

Comment:
We could have scenarios where the task succeeds, but we fail to reach this point to update the status, and then because it succeeded it fails the second time around. Having an idempotent move function could be seems quite valuable even for the initial version since it would be more forgiving.

Reply:
This has occured to me. Since there is no reliable way of running a task and the metadata update in the same transaction plus capture the output we now have this small window in which the task execution could be committed and the metadate update wouldn't.

I wonder how likely that is in normal circumstances.